app/gui/src/components/AppContainerLayout.vue

```diff
@@ -1,67 +1,94 @@
 import { backendQueryOptions } from '@/composables/backend'
 import { useEvent } from '@/composables/events'
+import { Ok } from '@/util/data/result'
 import { reactComponent } from '@/util/react'
 import { useQuery } from '@tanstack/vue-query'
-import { computed, onMounted, onUnmounted } from 'vue'
-import { Ok } from 'ydoc-shared/util/data/result'
+import { computed, onMounted, onUnmounted, proxyRefs } from 'vue'
 
 const SetupOrganizationModal = reactComponent(SetupOrganizationModalReact)
 const TrialEndedModal = reactComponent(TrialEndedModalReact)
+const PlanDowngradedModal = reactComponent(PlanDowngradedModalReact)
 
 const PLANS_TO_SPECIFY_ORG_NAME = [backendModule.Plan.team, backendModule.Plan.enterprise]
 
+type Props = {
+  shouldSetupOrganization: boolean
+  trialEndedModalProps?: TrialEndedModalProps | undefined
+  planDowngradedModalProps?: PlanDowngradedModalProps | undefined
+}
+
+/** Days of asset retention after trial ends. */
+const DAYS_BEFORE_DELETE = 90
+
 /**
  * A layout for "main app" router views.
  *
  * TODO[ao]: should be merged with `AppContainer` probably, but first we need to remove
  * the "Dashboard" layer between them.
  */
-export const dataLoader: DataLoader<{
-  shouldSetupOrganization?: boolean
-  showTrialEndedModal?: boolean
-}> = {
+export const dataLoader: DataLoader<Props> = {
   async beforeRouteEnter() {
     const auth = useAuth()
     const { remoteBackend: backend } = useBackends()
-    if (!auth.session) return Ok({})
-    const { isOrganizationAdmin, plan = backendModule.Plan.free } = auth.session.user
 
-    if (!isOrganizationAdmin || plan === backendModule.Plan.free) return Ok({})
+    const { isOrganizationAdmin, plan } = auth.session?.user ?? {
+      isOrganizationAdmin: false,
+      plan: backendModule.Plan.free,
+    }
+
+    const needsOrganizationSetup = PLANS_TO_SPECIFY_ORG_NAME.includes(plan)
 
     const organizationQuery = useQuery(backendQueryOptions('getOrganization', [], backend))
     await organizationQuery.suspense()
 
-    const trialEndedModalProps = computed(() =>
-      organizationQuery.data.value?.subscription?.isPaused ?
-        { subscriptionId: organizationQuery.data.value?.subscription?.id }
-      : undefined,
-    )
+    const trialEndedModalProps = computed<TrialEndedModalProps | undefined>(() => {
+      if (plan == backendModule.Plan.free) return undefined
 
-    if (!PLANS_TO_SPECIFY_ORG_NAME.includes(plan)) return Ok({ trialEndedModalProps })
+      const subscription = organizationQuery.data.value?.subscription
+      if (subscription?.isPaused && subscription.id != null) {
+        return { subscriptionId: subscription.id }
+      }
+    })
 
-    return Ok({
-      shouldSetupOrganization: computed(
-        () =>
-          organizationQuery.data.value?.name == null || organizationQuery.data.value?.name === '',
-      ),
-      trialEndedModalProps,
+    const planDowngradedModalProps = computed<PlanDowngradedModalProps | undefined>(() => {
```

Comment:
The modal-related code from this component could be moved to a separate composable.

Reply:
I don't think it easily can. This is part of the dataloader in the "layout" component, because the async call is supposed to block router navigation.